### ISSUE TITLE
Stop emitting `.debug_pubnames` and `.debug_pubtypes` into Linux object files and binaries

### ISSUE THREAD

**Issue by rocallahan** [rust-lang/rust#48762]
As far as I can tell, nothing uses these. They bloat executables and make linking more expensive.

At least stop emitting them for `debuginfo=1`.

**Comment by sfackler** [comment]
It seems like gcc and clang don't emit these anymore, so it'd probably make sense for us to skip them as well.

**Comment by tromey** [comment]
I have a patch for this but I'm still looking at how to write a test for it.

**Comment by tromey** [comment]
I learned on #llvm that these sections may be read by lld to create a `.gdb_index` section.  So perhaps they are desirable after all.  I'll look into lld to see.

**Comment by tromey** [comment]
`.debug_pubnames` and `.debug_pubtypes` are indeed read by `lld` when `--gdb-index` is given.  This doesn't appear to be the default.  This emits an index that speeds up gdb symbol reading.  So, it seems that Linux is the platform where these sections might be useful.  However, since it requires a linker flag, perhaps the thing to do is just drop this everywhere.

**Comment by DenialAdams** [comment]
> 
> 
> `.debug_pubnames` and `.debug_pubtypes` are indeed read by `lld` when `--gdb-index` is given. This doesn't appear to be the default. This emits an index that speeds up gdb symbol reading. So, it seems that Linux is the platform where these sections might be useful. However, since it requires a linker flag, perhaps the thing to do is just drop this everywhere.

It seems like we could skip emitting them by default, but add a cargo option that causes them to be emitted and turns the linker flag on?

**Comment by tromey** [comment]
> It seems like we could skip emitting them by default, but add a cargo option that causes them to be emitted and turns the linker flag on?

Yeah, that could be done.  I have other priorities, though, so I'm going to drop this for now.  I can share my rustc patch if anybody wants to pick it up.

**Comment by khuey** [comment]
What blocks landing this?
